pkg/driver/node/credentialprovider/provider.go

```diff
@@ -0,0 +1,125 @@
+// Package credentialprovider provides utilities for obtaining AWS credentials to use.
+// Depending on the configuration, it either uses Pod-level or Driver-level credentials.
+package credentialprovider
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/fs"
+	"strings"
+
+	k8sv1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	k8sstrings "k8s.io/utils/strings"
+
+	"github.com/awslabs/aws-s3-csi-driver/pkg/driver/node/envprovider"
+)
+
+// CredentialFilePerm is the default permissions to be used for credential files.
+// It's only readable and writeable by the owner.
+const CredentialFilePerm = fs.FileMode(0600)
+
+// CredentialDirPerm is the default permissions to be used for credential directories.
+// It's only readable, listable (execute bit), and writeable by the owner.
+const CredentialDirPerm = fs.FileMode(0700)
+
+// An AuthenticationSource represents the source where the credentials was obtained.
+type AuthenticationSource = string
+
+const (
+	// This is when users don't provide a `authenticationSource` option in their volume attributes.
+	// We're defaulting to `driver` in this case.
+	AuthenticationSourceUnspecified AuthenticationSource = ""
+	AuthenticationSourceDriver      AuthenticationSource = "driver"
+	AuthenticationSourcePod         AuthenticationSource = "pod"
+)
+
+// A Provider provides methods for accessing AWS credentials.
+type Provider struct {
+	client         k8sv1.CoreV1Interface
+	regionFromIMDS func() (string, error)
+}
+
+// A ProvideContext contains parameters needed to provide credentials for a volume mount.
+//
+// Here, [WritePath] and [EnvPath] are used together to provide credential files to Mountpoint.
+// The [Provider.Provide] method decides on filenames for credentials (e.g., `token` for driver-level service account token)
+// and writes credentials with these filenames in [WritePath], and returns environment variables to pass Mountpoint
+// with these filenames in [EnvPath].
+// This is due to fact that Mountpoint and the CSI Driver Node Pod - caller of this method - runs with different
+// filesystems, and to communicate with each other, the CSI Driver Node Pod uses `hostPath` volume to gain
+// access some path visible from both the CSI Driver Node Pod and Mountpoint, and setups files in that volume
+// using [WritePath] and returns paths to these files in [EnvPath], so Mountpoint can correctly read these files.
+type ProvideContext struct {
```

Comment:
`credentialprovider.Provide` method was accepting a few arguments including volume context before, now I grouped all them into `ProvideContext`.

The `WritePath` and `EnvPath` are the new ones, they were hard-coded before, and will be different on `PodMounter` and `SystemdMounter`.